src/dos/dos_code_page.cpp

```diff
@@ -0,0 +1,2223 @@
+	PrintWarning(font_offset, message);
+}
+
+void CpiParser::WarnFontSizeDuplicateOffsets(const uint32_t struct_offset,
+                                             const std::string& struct_name,
+                                             const uint8_t height)
+{
+	const auto message = format_str("multiple offsets found for screen font 8x%d",
+	                                height);
+	PrintWarning(struct_offset, struct_name, message);
+}
+
+void CpiParser::WarnUnknownScreenFontFormat(const uint32_t struct_offset,
+                                            const std::string& struct_name,
+                                            const uint16_t version)
+{
+	const auto message = format_str("unknown screen font format %d", version);
+	PrintWarning(struct_offset, struct_name, message);
+}
+
+void CpiParser::WarnScreenFontWithAspect(const uint32_t struct_offset,
+                                         const std::string& struct_name)
+{
+	PrintWarning(struct_offset,
+	             struct_name,
+	             "screen font has a non-zero aspect value");
+}
+
+void CpiParser::WarnScreenFontWidthInvalid(const uint32_t struct_offset,
+                                           const std::string& struct_name,
+                                           const uint8_t width)
+{
+	const auto message = format_str("screen font width %d is invalid", width);
+	PrintWarning(struct_offset, struct_name, message);
+}
+
+void CpiParser::WarnScreenFontDuplicateSize(const uint32_t struct_offset,
+                                            const std::string& struct_name,
+                                            const uint8_t height)
+{
+	const auto message = format_str("screen font size 8x%d found more than once",
+	                                height);
+	PrintWarning(struct_offset, struct_name, message);
+}
+
+void CpiParser::WarnScreenFontWithoutSizes(const uint32_t struct_offset,
+                                           const std::string& struct_name)
+{
+	PrintWarning(struct_offset, struct_name,
+	             "screen font does not define any font size");
+}
+
+void CpiParser::WarnScreenFontWithoutUsefulSizes(const uint32_t font_offset)
+{
+	PrintWarning(font_offset,
+	             "screen font does not contain any of the 8x16, 8x14 and 8x8 sizes");
+}
+
+void CpiParser::WarnScreenFontSizeWithoutCharacters(const uint32_t struct_offset,
+                                                    const std::string& struct_name)
+{
+	PrintWarning(struct_offset, struct_name,
+	             "screen font size does no contain any charcters");
+}
+
+void CpiParser::WarnUnknownScreenFontDevice(const uint32_t struct_offset,
+                                            const std::string& struct_name,
+                                            const std::string& device_name)
+{
+	const auto message = format_str("unknown screen font device '%s'",
+	                                device_name.c_str());
+	PrintWarning(struct_offset, struct_name, message);
+}
```

Comment:
Do you need all these methods? Can't you just construct the errors inline? IMO some minor duplication would be preferable to having so many different warning/error logging methods.

Reply:
OK, changed to generate part of the error strings inline (struct name and offset are still formed in one certain place to ensure consistency).

Exception: `WarnDataOutOfBounds`, as it is used in many places.